plugins/suspendmanager.cpp

```diff
@@ -0,0 +1,822 @@
+                if (!isSuitableAccess(npos))
+                    continue; // non walkable neighbour, not an exit
+
+                auto impassiblePlan = plannedImpassibleAt(npos);
+                if (!impassiblePlan)
+                    // walkable neighbour with no building scheduled, not in a dead end
+                    return;
+
+                if (leadsToDeadend.contains(impassiblePlan->id))
+                    continue; // already visited, not an exit
+
+                if (exit)
+                    return; // more than one exit, not in a dead end
+
+                exit = impassiblePlan;
+            }
+
+            if (!exit) return;
+
+            // exit is the single exit point of this corridor, suspend its construction job...
+            for (auto exit_job : exit->jobs) {
+                if (exit_job->job_type == df::job_type::ConstructBuilding) {
+                    suspensions[exit_job->id] = Reason::DEADEND;
+                }
+            }
+            // ...mark the current tile of the corridor as leading to a dead-end...
+            leadsToDeadend.insert(building->id);
+
+            // ...and continue the exploration from its position
+            building = exit;
+            pos = coord(building->centerx,building->centery,building->z);
+        }
+    }
+
+    void preserveDesigations (df::job* job){
+        CHECK_NULL_POINTER(job)
+        if (job->job_type == df::job_type::CarveTrack ||
+            job->job_type == df::job_type::SmoothFloor ||
+            job->job_type == df::job_type::DetailFloor)
+        {
+            auto building = Buildings::findAtTile(job->pos);
+            if (building) {
+                for (auto building_job : building->jobs) {
+                    if (building_job->job_type == df::job_type::ConstructBuilding) {
+                        suspensions[building_job->id] = Reason::ERASE_DESIGNATION;
+                    }
+                }
+            }
+        }
+    }
+
+    // try to fetch suspension reason for a job without modifying the map
+    // (std::map::operator[] inserts the key if it isn't found)
+    bool tryGetReason (df::job* job, Reason &reason) {
+        auto reason_it = suspensions.find(job->id);
+        if (reason_it != suspensions.end()) {
+                reason = reason_it->second;
+                return true;
+        }
+        return false;
+    }
+
+    std::map<int,Reason> suspensions;
+    std::set<int> leadsToDeadend;
+    size_t num_suspend = 0, num_unsuspend = 0;
+
+public:
+    bool prevent_blocking = true;
+
+    // gather some statistics about the last call to do_cycle()
+    string getStatus (color_ostream &out) {
+        std::map<Reason,int> stats;
+
+        for (auto [_ , reason] : suspensions) {
+            stats[reason == Reason::DEADEND ? Reason::RISK_BLOCKING : reason]++;
+        }
+
+        std::stringstream res;
+        res << "maintaining " << suspensions.size() << " suspension reasons\n";
```

Comment:
how about "%zd jobs suspended"?

Reply:
yes, noticed after the push that I accidentally deleted that line just before pushing. Will reinstate that line with the next push. :sweat_smile: